File: test/envelope.test.ts
import { describe, expect, it } from "vitest";
import vectors from "../spec/envelope-vectors.json";
import {
  type EnvelopeMessage,
  type FullEnvelopeMessage,
  renderEnvelope,
  renderFull,
} from "../src/lib/transit/envelope";

describe("transit/1 envelope", () => {
  for (const vector of vectors.envelopes) {
    it(vector.name, () => {
      expect(renderEnvelope(vector.input as EnvelopeMessage)).toBe(vector.expected);
    });
  }

  for (const vector of vectors.full) {
    it(vector.name, () => {
      expect(renderFull(vector.input as FullEnvelopeMessage)).toBe(vector.expected);
    });
  }

  it("clips direct-message injection to 4,000 runes", () => {
    const rendered = renderEnvelope({
      from: "alice@alpha",
      id: "tx_56789abcdef0",
      ts: "2026-08-21T12:35:01.000Z",
      kind: "dm",
      body: `${"🙂".repeat(4_000)}x`,
    });
    expect(rendered).toContain('truncated="1" schema="transit/1"');
    expect(rendered).toContain(`\n${"🙂".repeat(4_000)}\n`);
    expect(rendered).not.toContain(`${"🙂".repeat(4_000)}x`);
  });

  it("limits a channel preview to 100 runes", () => {
    const rendered = renderEnvelope({
      from: "telegram",
      id: "dlv_6789abcdef01",
      ts: "2026-08-21T12:35:02.000Z",
      kind: "channel",
      body: "🙂".repeat(101),
      conversationId: "chat-2",
      connector: "telegram",
    });
    expect(rendered).toContain(`\n${"🙂".repeat(100)}\n`);
    expect(rendered).not.toContain("🙂".repeat(101));
  });
});

File: daemon/envelope_test.go
package main

import (
	"encoding/json"
	"os"
	"strings"
	"testing"
)

func TestEnvelopeVectors(t *testing.T) {
	data, err := os.ReadFile("../spec/envelope-vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var vectors struct {
		Envelopes []struct {
			Name     string        `json:"name"`
			Input    EnvelopeInput `json:"input"`
			Expected string        `json:"expected"`
		} `json:"envelopes"`
		Full []struct {
			Name     string            `json:"name"`
			Input    FullEnvelopeInput `json:"input"`
			Expected string            `json:"expected"`
		} `json:"full"`
	}
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatal(err)
	}
	for _, vector := range vectors.Envelopes {
		t.Run(vector.Name, func(t *testing.T) {
			if got := RenderEnvelope(vector.Input); got != vector.Expected {
				t.Fatalf("RenderEnvelope()\n--- got ---\n%s\n--- want ---\n%s", got, vector.Expected)
			}
		})
	}
	for _, vector := range vectors.Full {
		t.Run(vector.Name, func(t *testing.T) {
			if got := RenderFull(vector.Input); got != vector.Expected {
				t.Fatalf("RenderFull()\n--- got ---\n%s\n--- want ---\n%s", got, vector.Expected)
			}
		})
	}
}

func TestEnvelopeClipsByRune(t *testing.T) {
	body := strings.Repeat("🙂", 4000) + "x"
	rendered := RenderEnvelope(EnvelopeInput{
		From: "alice@alpha", ID: "tx_56789abcdef0", TS: "2026-08-21T12:35:01Z",
		Kind: "dm", Body: body,
	})
	if !strings.Contains(rendered, `truncated="1" schema="transit/1"`) {
		t.Fatalf("missing truncation attribute: %s", rendered[:200])
	}
	if strings.Contains(rendered, body) {
		t.Fatal("unclipped body present")
	}
}
